keytree/src/lib.rs

```diff
@@ -37,6 +41,41 @@ impl Xprv {
     }
 }
 
+/// Returns a intermediate child pubkey.
+pub fn intermediate_key(
+    xpub: &Xpub,
+    label: &'static std::string::String,
+    data: &String,
+) -> Option<Xpub> {
+    // question: this function returns an option because it calls a function that returns an option
+    // and I didn't know what else to do. is there something better to do?
+
+    // question: is the following kosher? https://doc.dalek.rs/merlin/struct.Transcript.html#note
+    let mut t = Transcript::new(b"Keytree.intermediate");
+    t.commit_bytes(b"pt", xpub.point.as_bytes());
+    t.commit_bytes(b"dk", &xpub.dk);
+
+    // provide the transcript to the user to commit an arbitrary derivation path or index
+    t.commit_bytes(label.as_bytes(), data.as_bytes());
+
+    // squeeze a challenge scalar
+    let f = t.challenge_scalar(label.as_bytes()); // is this the right input?
+
+    // squeeze a new derivation key
+    let mut dk = [0u8; 32];
+    t.challenge_bytes(b"dk", &mut dk);
+
+    let point;
+    match xpub.point.decompress() {
```

Comment:
We can avoid failable decoding by pushing the decoding up the API stack:

1. Xpub can hold decompressed `RistrettoPoint`.
2. Xprv->Xpub conversion does not fail.
3. bytes -> Xpub may fail because it will have to decompress the point.
4. To avoid overhead of recompressing the point, Xpub could also _own_ the transcript instance precomputed with `pt` and `dk`. Then, when we need to derive things, we'd clone the transcript, customize it with user's data and then squeeze the factor. When converting 64 bytes into Xpub, this precomputation would happen with an already compressed point (which is == the first 32 bytes).
5. To reuse the transcript for both intermediate and leaf cases, I suggest changing the spec: the label will be the same for both uses: `"Keytree.derive"`, but domain separation will happen when we squeeze `f`: `"f.intermediate"`  and `"f.leaf"` respectively.
6. As a result, derivation functions will be non-failable.

Reply:
Private function `Xpub::from_parts(RistrettoPoint, [u8;32])` can be used by Xprv->Xpub conversion and by derivation function.

Reply:
I have done the very first part of this at https://github.com/interstellar/slingshot/pull/207

Reply:
Sorry I've been going through this so slowly. Another question: 

are you suggesting that we add another field to the xpub for the transcript, and when the xpub is generated from the xprv, we also produce a new transcript instance using that xpub’s point and derivation key?

Reply:
>are you suggesting that we add another field to the xpub for the transcript, 

Yes. `Xpub {point, dk, transcript}`

>and when the xpub is generated from the xprv, we also produce a new transcript instance using that xpub’s point and derivation key?

Not quite: simply clone xprv.transcript to assign it to xpub.transcript.

Reply:
^ this is because the transcript holds the derivation that's the same for xpubs and xprvs, so it must be exactly the same in xpub and xprv

Reply:
Ah, ok, so we also store the transcript on Xprv? And we create that instance of the transcript when we create the Xprv?

Reply:
yup!

Reply:
And it's supposed to be precomputed:

```
t = Transcript::new()
t.commit(pub)
t.commit(dk)
Xpub.transcript = t
```

Reply:
so that when you derive a child key, you can simply do:

```
t = self.transcript.clone()
t.commit(user provided derivation path)
f = t.challenge_scalar("f.leaf") // or "f.intermediate"
...
```

Reply:
In this pseudo code: 

```
t = Transcript::new()
t.commit(pub)
t.commit(dk)
Xpub.transcript = t
```

Did you mean to set the transcript on the Xprv instead? And if so, what are you committing? Finally, I assume you need to use `commit_bytes` to commit things to the transcript.... unless we should do the closure trick you mentioned above ^ ?

Reply:
sorry for not being clear. 

When you initialize _either_ Xpub, or Xprv, they both need to instantiate the same precomputed transcript.

* the `commit(pub)` is actually `commit_point("pt", point.compress())`, point is computed on the fly for Xprv (from `scalar*B`) and for Xpub it's already there.
* the `commit(dk)` is actually `commit_bytes("dk", &dk)`

Then, the transcript is stored inside the structure (Xpub or Xprv).

When you derive Xpub from Xprv, the transcript is already good, you just need to compute point `xpub.point = xprv.scalar*B` and clone the transcript.

Reply:
edited ^

Reply:
When you derive a child, you use a closure on Xprv/Xpub to which you clone `t=self.transcript.clone()` and provide `&mut t` to let user customize it with derivation path.

After that, you squeeze `f` and (if it's an intermediate key) `dk2` from that transcript. Then, you instantiate a new (child) Xprv/Xpub and precompute its own transcript from scratch using its pubkey (adjusted by `f`) and `dk2`.